base/compiler/tfuncs.jl

```diff
@@ -1568,7 +1570,7 @@ function builtin_tfunction(interp::AbstractInterpreter, @nospecialize(f), argtyp
     end
     if isa(f, IntrinsicFunction)
         if is_pure_intrinsic_infer(f) && _all(@nospecialize(a) -> isa(a, Const), argtypes)
-            argvals = anymap(a::Const -> a.val, argtypes)
```

Comment:
Aren't these basically exactly equivalent?

Reply:
Not really, if the call signature is restricted the inlinear needs to insert a branch for dynamic call that accounts for possible `MethodError` (, which in turn makes it hard to inline the expanded body):
```julia
julia> code_typed((Any,Any)) do x, y
           xx = (a::Core.Const -> a.val)(x)                  # not inlineable
           yy = (@nospecialize(a) -> (a::Core.Const).val)(y) # inlineable
           xx, yy
       end
1-element Vector{Any}:
 CodeInfo(
1 ─ %1  = %new(Base.:(var"#944#947"))::Base.var"#944#947"
│   %2  = (isa)(x, Core.Const)::Bool
└──       goto #3 if not %2
2 ─ %4  = π (x, Core.Const)
│   %5  = Base.getfield(%4, :val)::Any
└──       goto #4
3 ─ %7  = (%1)(x)::Any
└──       goto #4
4 ┄ %9  = φ (#2 => %5, #3 => %7)::Any
│   %10 = Core.Const::Type{Core.Const}
│         Core.typeassert(y, %10)::Core.Const
│   %12 = π (y, Core.Const)
│   %13 = Base.getfield(%12, :val)::Any
│   %14 = Core.tuple(%9, %13)::Tuple{Any, Any}
└──       return %14
) => Tuple{Any, Any})
```